components/remerge/src/vclock.rs

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Our implementation of vector clocks. See the remerge RFC's appendix for an
+//! overview of how these work if you're unfamiliar.
+
+use rusqlite::types::{FromSql, FromSqlError, FromSqlResult, ToSql, ToSqlOutput, ValueRef};
+use std::collections::BTreeMap;
+use sync_guid::Guid;
+
+pub type Counter = u64;
+
+/// A vector clock.
+#[derive(Clone, Default, Debug, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
+pub struct VClock(pub BTreeMap<Guid, Counter>);
+
+/// Basically equivalent to Option<std::cmp::Ordering>, but more explicit about
+/// what each value means. The variant documentation assumes this is generated by
+/// something similar to `lhs.get_ordering(rhs)`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ClockOrdering {
+    /// The two clocks are equivalent.
+    Equivalent,
+    /// The `lhs` clock is an ancestor of the `rhs` clock.
+    Ancestor,
+    /// The `lhs` clock is a decendent of the `rhs` clock.
+    Descendent,
+    /// The two clocks are in conflict, and some other means of resolution must
+    /// be used.
+    Conflicting,
+}
+
+impl VClock {
+    pub fn new(own_client_id: Guid, counter: Counter) -> Self {
+        VClock(std::iter::once((own_client_id, counter)).collect())
+    }
+
+    /// Determine the ordering between `self` and `other`.
+    pub fn get_ordering(&self, other: &VClock) -> ClockOrdering {
+        let mut seen_gt = false;
+        let mut seen_lt = false;
+
+        let self_kvs = self.0.iter().map(|(id, &ctr)| (id, Some(ctr), None));
+        let other_kvs = other.0.iter().map(|(id, &ctr)| (id, None, Some(ctr)));
+
+        for (k, sv, ov) in self_kvs.chain(other_kvs) {
+            let sv = sv.unwrap_or_else(|| self.get(k));
+            let ov = ov.unwrap_or_else(|| other.get(k));
```

Comment:
This seems complicated. I would have probably written

```
for k in self.0.keys().chain(other.0.keys()) {
    let sv = self.get(k);
    let ov = other.get(k);
```

Maybe your version is more efficient? Is that the motivation?

Reply:
I think your solution leads to exactly one value lookup for every iteration through the loop, and my solution leads to exactly two value lookups for every iteration through the loop (but, potentially with fewer loops).

Reply:
No, you're right, your solution definitely is more efficient.